### ISSUE TITLE
Duplicated artifacts in cache

### ISSUE THREAD

**Issue by tilusnet** [python-poetry/poetry#5921]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [ ] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**:  Linux Ubuntu 20.04
- **Poetry version**: 1.1.13

## Issue

Poetry was able somehow to create multiple same name entries in the cache:

```
~/.cache/pypoetry/artifacts$ find . -name forma*
./f7/4a/92/77326137086392f59eb66c784e369a6ab387f4178c4b950abe48ed5c4c/formasaurus-0.9.0+ds.3-py2.py3-none-any.whl
./e5/21/35/b92d6ab0cf36e72fc6a56d5ae502caf69b52f953a671d9e196cb1b4abd/formasaurus-0.9.0+ds.2-py2.py3-none-any.whl
./c4/38/b9/6d7ce39991c8319ed8ac62765317620a34d76f60f30152ca89e85d298b/formasaurus-0.9.0+ds.2-py2.py3-none-any.whl
./d9/25/49/b44cc41365aeb2139a3a5a87d530560a01faf827479ed1cffbcb6c2e0b/formasaurus-0.9.0+ds.3-py2.py3-none-any.whl
```
This will likely generate broken `poetry.lock` files since the collected artifact (by name) is no longer deterministic.

**Comment by dimbleby** [comment]
https://github.com/python-poetry/poetry/blob/ddcf9489fc251c6abd21570b326e007f2434eacc/src/poetry/installation/chef.py#L72 is the code that calculates the hash directory for a link.  

It should be entirely deterministic.  Presumably one of the inputs is changing each time it is generating a new directory.

> This will likely generate broken poetry.lock files ... 

Do you have an actual bug here or only a worry that something might have gone wrong?

I don't think that these files are evidence that anything has gone wrong, but if you have a real bug please say so.

**Comment by tilusnet** [comment]
I did end up with broken lock files, yes.
Please see my cache listing: it has two duplicates which should never happen imo.

**Comment by dimbleby** [comment]
What does "broken lock files" mean?  Can you share a "broken" lock file?

The cache listing that you have presented is not in itself an issue.  

Please say what your actual problem is.  What is it that you are trying to do, and how is it failing?

**Comment by tilusnet** [comment]
A broken lock file is one that once built and shared cannot be used elsewhere apart from the system on which was built. 

In the scenario above what can happen (and happened):
1. the local system runs `poetry update` picking one of the cached artifact - let's call this variant A of formasaurus-0.9.0+ds.3
2. a remote system receives this lockfile but it can only resolve to variant B of formasaurus-0.9.0+ds.3
3. the build will fail on the remote system.

Please take note that artifacts built on Dev systems may occasionally have the same name and version but different binary content as part of the Dev cycle. For this reason alone it is instrumental that your tool is not storing several variants of the same named artifact under its cache.

**Comment by dimbleby** [comment]
If you are:

- producing a lock file pinning a package with sha 1
- but your server is returning a version of that package with sha 2

then it is expected that poetry install will fail due to hash mismatch.

This is nothing to do with the local cache.  Indeed the existence of multiple files suggests that poetry is successfully recognising that these are different binaries (per the code I linked, the hash returned by the server is incorporated into the directory path, so this is probably what is happening).

**Comment by tilusnet** [comment]
I am not pinning to a sha! I am referencing a dependency with (name, version). When I build, as a dev that's what I care about. 
I will never be interested in poetry picking up a stale version from the cache.

Put it this way, if the poetry cache contains both
- (package name A, package version 1, sha X)
- (package name A, package version 1, sha Y)

and a project lists as dependency (package name A, package version 1), which one is poetry "designed to" pick up from the cache?

This is why poetry must always keep a single variant of an artifact as designated by (name, version) parameters.
When an updated (name, version) artifact is detected by poetry, the old one in cache must go away / overwritten by the new one.

Please fix this problem.

**Comment by tilusnet** [comment]
Yeah I believe you still didn't understand the severity of this issue.

Here is a typical scenario, step by step.

Objectives: 
* Build Project A with dependency B. 
* Both project A and dependency B are of our own, we build them both 
* Dependency B is to be published in our own repository. 

Steps:
On the local dev machine:
1. Build dependency B under version 1 (-> wheel file produced)
2. Build project A using dependency B version 1 (-> poetry registers depB,ver1 in local cache) 
3. Rebuild dependency B under version 1 (-> updated wheel file produced)
4. Rebuild project A using dependency B version 1 (-> poetry registers a _new_ depB,ver1 in local cache) 
5. Publish final dependency B, version 1 to internal repository (notice: there is a unique artifact in the repo for depB,ver1 -- the latest build)
6. Publish the poetry.lock file for remote CI execution (-> poetry lock file has now a sha reference that is undetermistic due to the duplicated artifacts in the poetry cache)

On the remote CI machine:
1. poetry lock file for Project A gets pulled
2. referred artifact dependency B, version 1 is found in internal repository
3. the build fails if the poetry lock file happened to pick the mismatching sha artifact from the local dev machine's cache 

I hope now it is clear why duplicated artifacts are a problem in a local poetry cache.

**Comment by dimbleby** [comment]
> 2.	Build project A using dependency B version 1 

In this step, how is A referring to B?  Is it as a path dependency, or is there already a B1 uploaded to your internal repository, or what?

If there's already a B1 uploaded to your repository then see my last comment: you are treating artifacts as mutable, it is expected that this will cause hashes to become outdated, recommend don't do that.

If as a path dependency then I wouldn't expect this to cause anything to appear in the local cache

**Comment by tilusnet** [comment]
What is the rationale behind storing several variants of the same artifact version inside the cache?
What kind of good purpose does it serve?

**Comment by dimbleby** [comment]
As in:
```toml
b = { path = "/path/to/b" }
```
where this is the directory that holds B's pyproject.toml?

I cannot reproduce this putting anything into the local cache.

**Comment by dimbleby** [comment]
> What is the rationale behind storing several variants of the same artifact version inside the cache?

See the code that I pointed to at the start of this thread

**Comment by tilusnet** [comment]
rationale = a set of reasons or a logical basis for a course of action or belief.

in simpler terms: why did you implement it that way?

**Comment by dimbleby** [comment]
I didn't implement this code at all.

Can you confirm whether you are or are not overwriting binary artifacts on your server at any point in your workflow?

**Comment by dimbleby** [comment]
which part of it seems unreasonable?  I'd have doubts about including the interpreter name and version myself, but I see no hint that this is relevant in this discussion.

So far as I can tell, path dependencies don't result in anything being added to the cache at all; so we still don't have a shared understanding of what you are actually doing.

Perhaps you could share your pyproject.toml, or put together a small project by which someone else can reproduce what you are doing.

**Comment by tilusnet** [comment]
This part is unreasonable:

if the poetry cache is able to (designed to?) store multiple binaries under the same version
* (package name A, package version 1, sha X)
* (package name A, package version 1, sha Y)

how is it deciding which one it resolves to when a project refers to Package A, version 1?

**Comment by dimbleby** [comment]
By looking in the unique directory for the `Link`, as defined by the code that I pointed at

**Comment by tilusnet** [comment]
I meant _in principle_. By design concept.

It's irrelevant what is in the code. I could toss a coin and implement it -- that doesn't mean it makes any practical sense.

**Comment by dimbleby** [comment]
It feels as though this conversation is becoming unproductive.  Perhaps others will have better luck understanding what you are doing.

> So far as I can tell, path dependencies don't result in anything being added to the cache at all; so we still don't have a shared understanding of what you are actually doing.
>
> Perhaps you could share your pyproject.toml, or put together a small project by which someone else can reproduce what you are doing.

If you can provide more details per above, perhaps that will help someone to help you.

**Comment by tilusnet** [comment]
I agree, I think you are unable to address my concerns despite attempting to.

You seem to refer back repeatedly to the existing codebase without being able to explain why the implementation is the way it is.
It is that very implementation I am challenging here. 

Because, as I tried to illustrated above, ending up with several artifacts under the same package version opens up a can of worms leading to hard to understand / undeterministic behaviour. It is the poetry tooling that put those duplicates in my cache, I haven't put them there myself. In that sense the tool has an issue.

By all means thank you @dimbleby for your time, but I think resolution needs a different tack here.

**Comment by finswimmer** [comment]
I cannot help here either, because I have no idea how the caching works.

I have taken a look into my cache and find there also multiple locations for the same whl file, e.g.:

```
./39/82/d0/88d2e96d51ddcee6b8edac30d02b0a99dc19a34d42cfca6eecd05f25ce/xsdata-22.5-py3-none-any.whl
./7e/ec/81/b48225c1051a05e6d042a08b59006d2e9c3e6baf5f6ee4ad1f2b39f598/xsdata-22.5-py3-none-any.whl
```

But the hash you are seeing in the path, seems to have nothing to do with the hash of the file:

```
sha256sum ./39/82/d0/88d2e96d51ddcee6b8edac30d02b0a99dc19a34d42cfca6eecd05f25ce/xsdata-22.5-py3-none-any.whl                                                                                                  
817f73bd0374305ebbef2de1a7c93b7ef8bfdeaf053d057436b037e026fad535  ./39/82/d0/88d2e96d51ddcee6b8edac30d02b0a99dc19a34d42cfca6eecd05f25ce/xsdata-22.5-py3-none-any.whl
```

```
sha256sum ./7e/ec/81/b48225c1051a05e6d042a08b59006d2e9c3e6baf5f6ee4ad1f2b39f598/xsdata-22.5-py3-none-any.whl          
817f73bd0374305ebbef2de1a7c93b7ef8bfdeaf053d057436b037e026fad535  ./7e/ec/81/b48225c1051a05e6d042a08b59006d2e9c3e6baf5f6ee4ad1f2b39f598/xsdata-22.5-py3-none-any.whl
```

As you can see they are identical. And they have the same value as in my `poetry.lock` file:

```
xsdata = [
    {file = "xsdata-22.5-py3-none-any.whl", hash = "sha256:817f73bd0374305ebbef2de1a7c93b7ef8bfdeaf053d057436b037e026fad535"},
    {file = "xsdata-22.5.tar.gz", hash = "sha256:701dcda5a4a07df3a642f8fb6f73f0e2e18224d359bcb7f4212b3c29d7af72c2"},
]
```

**Comment by dimbleby** [comment]
Anyway so far as I can see this cache is not used at all during locking - only during install.  So - as best I can tell what OP's issue is, which is not very well at all - the cache is not relevant to it.

**Comment by tilusnet** [comment]
@dimbleby, I can confirm though that once I ran `poetry cache clear --all`, and reran `poetry update`, the regenerated lock file got the sha corrected in relation to dependency B. So in a way the cache _does_ have to do with lock file generation during `poetry update`, either directly or indirectly.

Regarding testing in your environment, did you also test by pointing at the wheel file of the dependency? That's what I typically do. Some of our dependencies are _not_ poetry managed, hence they don't have a pyproject.toml. We build them with plain old `python setup.py bdist_wheel`. 

```
dependency-B = { path = "/path/to/a.whl", develop = false }
```

@finswimmer thanks for your input. It's interesting to see that you also got duplicates for same version artifacts.
The artifact path is indeed unrelated to the sha of the artifact itself - and indeed, it should not be.
If the interpreter version influences the cache path, that is understandable to me -- it is a version dependendency after all and the cache may prefer to differentiate on that. I think that's a sound design choice.
Though I reckon your `xsdata-22.5` is always coming directly from a public repository? If so, that explains why the file sha of all your cached copies is identical -- those are copies of the same binary.

**Comment by tilusnet** [comment]
no, CI / deployment systems always use named dependencies.
path dependencies are for local development only.

please see an exact recipe of what we're doing as outlined above https://github.com/python-poetry/poetry/issues/5921#issuecomment-1168475400

**Comment by dimbleby** [comment]
That is very far from being an exact recipe.  Even now you are hinting at - but not spelling out things in your process which that comment does not mention

- path dependency points at whl
- some sort of change from path to named dependency for CI

Even with a dependency like
```toml
dependency-B = { path = "/path/to/a.whl", develop = false }
```
I am unable to get `a.whl` (I suppose you meant `b.whl`, but either way) to appear in this cache.

This thread is getting on for 30 comments deep and you are still the only person who knows what you are doing.  Whether I am not reading clearly or you are not writing clearly or both or neither: it's not working.  Which is why I suggest that an executable repro may be the way to go.

**Comment by finswimmer** [comment]
Hey @tilusnet,

I will close this as there seems to be no issue and Poetry behaves like expected (https://github.com/python-poetry/poetry/issues/5921#issuecomment-1168593540, https://github.com/python-poetry/poetry/issues/5921#issuecomment-1168602395)

Feel free to leave more information if you think there's still an issue.

fin swimmer
